pyrelational/informativeness/abstract_scorers.py

```diff
@@ -0,0 +1,78 @@
+"""Abstract scorer classes to define the interface for informativeness scorers."""
+
+from abc import ABC, abstractmethod
+from typing import Any, Union
+
+from torch import Tensor
+from torch.distributions import Distribution
+
+from .decorators import require_2d_tensor, require_probabilities
+
+
+class AbstractScorer(ABC):
+    """Abstract scorer class."""
+
+    @abstractmethod
+    def __call__(self, *args: Any, **kwargs: Any) -> Tensor:
+        """Score the input data.
+
+        This method should be implemented by the subclasses.
+        :return: scores associated with the input data.
+        """
+        pass
+
+
+class DecoratedClassificationScorerMeta(type):
+    """Metaclass for classification scorers."""
+
+    def __new__(cls, name: str, bases: tuple[Any, ...], dct: dict[str, Any]) -> "DecoratedClassificationScorerMeta":
+        """Decorate the `__call__` method with the `require_2d_tensor` and `require_probabilities` decorators."""
+        for key, value in dct.items():
+            if callable(value) and key == "__call__":
+                dct[key] = require_probabilities(require_2d_tensor(value))
```

Comment:
Sorry I think I need a bit of hand-holding to get through this part. Why do we need informativeness scores to be 2D probability distributions? What is the role of the dictionary?

Reply:
So 2D requirements is actually breaking the repo, so I removed it. But something we should think about though. In classification, it looks like most scoring methods expects batch by number of classes matrix input, the third dimensions which would be for instance number of estimators is not clearly used. We also have that .mean(0) always in ClassificationStrategy post train_and_infer call that collapses the estimator dimension in all cases (that probably should be removed tbh, putting the responsibility on the user to do that rather that making such an assumptions ourselves, I keep it here because it's not the object of this PR but something to keep in mind).

The Metaclass is used here to automatically decorate the __call__ methods of any child classes instead of having to redecorate the method each time. Maybe something to relax if there's a score that doesn't require probabilities.